### ISSUE TITLE
[hipSYCL Warning] backend_loader

### ISSUE THREAD

**Issue by misos1** [AdaptiveCpp/AdaptiveCpp#671]
When I compile with `--hipsycl-targets=hip:gfx900` I am getting warning:

```
[hipSYCL Warning] backend_loader: Could not load backend plugin: /opt/hipSYCL/bin/../lib/hipSYCL/librt-backend-cuda.so
[hipSYCL Warning] libcuda.so.1: cannot open shared object file: No such file or directory
```

With just `--hipsycl-targets=omp` there is additional warning:

```
[hipSYCL Warning] backend_loader: Could not load backend plugin: /opt/hipSYCL/bin/../lib/hipSYCL/librt-backend-hip.so
[hipSYCL Warning] /opt/hipSYCL/bin/../lib/hipSYCL/librt-backend-hip.so: undefined symbol: hipDriverGetVersion
```

I suppose with just `omp` it was not linked with rocm hip libraries.

So I suppose that it is trying to load each possible backend regardless of whether it needs it or not. Maybe it would be better if the resulting binary loaded just the backends it really needs so for example with `--hipsycl-targets=hip:gfx900` would be loaded just the hip backend and so on.

Also I noticed that packages `hipsycl-base-nightly` and `hipsycl-base-rocm-nightly` are really big. Seems `hipsycl-base-nightly` contains llvm and `hipsycl-base-rocm-nightly` contains rocm. Would it not be possible to use hipsycl with llvm and rocm versions already installed on the system? I suppose this can be customised in syclcc.json. But mentioned packages are probably unavoidable. (Maybe they could be declared as `Recommends` instead of `Depends`? Or `hipsycl-rocm-nightly` could be just satisfied with the official rocm-hip package. But I understand that it is maybe easier to make hipsycl compatible and well-tested with some specific llvm and rocm version.)

Also I tried to compile example from here: https://sycl.readthedocs.io/en/latest/iface/usm_allocator.html
I needed to change `C[idx] = A[idx] + B[idx];` into `C[idx[0]] = A[idx[0]] + B[idx[0]];`. Is `sycl::id` from hipsycl somehow not compatible with the official `sycl::id`?

**Comment by misos1** [comment]
Btw it outputs errors with `--hipsycl-targets=spirv`:

```
clang-11: error: unknown argument: '-fsycl-device-only'
clang-11: error: unknown argument: '-fsycl-unnamed-lambda'
clang-11: error: unknown argument: '-fno-sycl-use-bitcode'
```

**Comment by illuhad** [comment]
> I suppose with just omp it was not linked with rocm hip libraries.

I think the issue is that the HIP plugin seems to not be linked to all necessary ROCm libraries as it should. Those then only get linked when using `--hipsycl-targets=hip:...`. This should not be necessary.

> Maybe it would be better if the resulting binary loaded just the backends it really needs so for example with --hipsycl-targets=hip:gfx900 would be loaded just the hip backend and so on.

I think we have to distinguish between backend support for the runtime and for which devices kernels are compiled. There can be use cases where you don't wish to compile kernels for a particular device, but still want to have runtime support - for example if you just want to use interop mechanisms to use e.g. cuBLAS which contains already precompiled kernels.
Therefore, I believe the current behavior of loading all runtime plugins is correct. This also then gives the application the correct view regarding device topology.
Note that you can explicitly prevent runtime plugins from loading using the `HIPSYCL_VISIBILITY_MASK` environment variable.

> Also I noticed that packages hipsycl-base-nightly and hipsycl-base-rocm-nightly are really big. Seems hipsycl-base-nightly contains llvm and hipsycl-base-rocm-nightly contains rocm. Would it not be possible to use hipsycl with llvm and rocm versions already installed on the system?

In this case the easiest solution is to compile hipSYCL from source, which is usually straight-forward. There are so many different ways in which distributions install LLVM and also ROCm installations may vary greatly depending on version that we currently don't want to support all conceivable combinations using prebuilt binaries.

> Also I tried to compile example from here: https://sycl.readthedocs.io/en/latest/iface/usm_allocator.html
I needed to change C[idx] = A[idx] + B[idx]; into C[idx[0]] = A[idx[0]] + B[idx[0]];. Is sycl::id from hipsycl somehow not compatible with the official sycl::id?

SYCL 2020 has added an implicit conversion operator from `id<1>` to `size_t`. We don't have this one yet. We should fix it.

> Btw it outputs errors with --hipsycl-targets=spirv:

As outlined in the documentation, SPIR-V requires a clang that has patches for SYCL support and SPIR-V code generation merged. clang 11 is far too old for this. Also, SPIR-V support is highly experimental, might only work with specific versions of those patches and some functionality is still unimplemented.

**Comment by misos1** [comment]
Did you consider using some reasonable default for `--hipsycl-targets` based on actually available hardware?  For example if there is vega 10 and openmp is available then it could be set to `omp;hip:gfx900`.

**Comment by illuhad** [comment]
> Did you consider using some reasonable default for --hipsycl-targets based on actually available hardware? For example if there is vega 10 and openmp is available then it could be set to omp;hip:gfx900.

It's worth some thought, however for many use cases in HPC the system on which compilation takes place is not the same one where the application is executed. I suspect finding out the available hardware configuration can be cumbersome to get to work in a way that works reliably across ROCm/CUDA versions.
You can set a default in `syclcc.json`, although we should probably expose this in a better way through cmake.

> Also I tried to compile example from here: https://sycl.readthedocs.io/en/latest/iface/usm_allocator.html
I needed to change C[idx] = A[idx] + B[idx]; into C[idx[0]] = A[idx[0]] + B[idx[0]];. Is sycl::id from hipsycl somehow not compatible with the official sycl::id?

This should now work with PR #673

**Comment by misos1** [comment]
> It's worth some thought, however for many use cases in HPC the system on which compilation takes place is not the same one where the application is executed.

Yes, but it seems also hipcc and cuda have some default. (Maybe because then it is simple to just type `hipcc hello.cpp` to compile some "hello world" program?)

> I suspect finding out the available hardware configuration can be cumbersome to get to work in a way that works reliably across ROCm/CUDA versions.

Maybe there are also other possibilities:

1. Use all possible targets for all available backends.
2. Use some older target for each backend which is backward compatible with newest hardware (but that may be a problem).
3. Just use `omp` (maybe with some warning about performance like with missing -O flag).

**Comment by illuhad** [comment]
> Yes, but it seems also hipcc and cuda have some default. (Maybe because then it is simple to just type hipcc hello.cpp to compile some "hello world" program?)

As these tools are part of their respective platforms, they have it easier to extract this information and take into account changes in CUDA/ROCm.

> Maybe there are also other possibilities:
>    Use all possible targets for all available backends.
 >   Use some older target for each backend which is backward compatible with newest hardware (but that may be a problem).
 >   Just use omp (maybe with some warning about performance like with missing -O flag).

Can you provide some more background about your use case, and why this feature is important to you? 

Generally my philosophy is that I don't want to encourage anti-patterns that might negatively affect performance. I think users should spend a second to think about what targets they want to optimize for, as it can have significant performance impact. As such, I'm hesitant to just make it use some old target by default that happens to work everywhere, which then causes suboptimal performance. This might especially cause surprises for users who are new to hipSYCL.

One thing that makes a lot of sense in my mind is to provide a `-DDEFAULT_TARGETS` cmake option. Users who don't wish to specify the targets again could then just configure hipSYCL at build time with a default set, which `syclcc` would then use by default.
We already have something that goes in that direction, but only for the (deprecated) target platform, not for the current targets specification.

**Comment by misos1** [comment]
> Can you provide some more background about your use case, and why this feature is important to you?

No, this is not really important. (Btw seems cuda uses some old default "sm_52".)

> One thing that makes a lot of sense in my mind is to provide a `-DDEFAULT_TARGETS` cmake option.

That looks good.
